stable/gocd/templates/gocd-agent-deployment.yaml

```diff
@@ -22,6 +22,11 @@ spec:
         release: {{ .Release.Name | quote }}
         component: agent
     spec:
+      {{- if .Values.agent.serviceAccountName }}
+      serviceAccountName: {{ .Values.agent.serviceAccountName }}
+      {{- else if .Values.serviceAccount.create }}
```

Comment:
@MPV Got back to this. Talked to @sheroy too.

This part seems to be problematic. As @varshavaradarajan was saying earlier, this makes the static agents have the same service account (newly created one) that the server uses. This means, any normal `helm install` with `agent.replicaCount` set to a non-zero value will make agent pods have a service account attached to them, which allows them to make API calls to k8s. This seems to be insecure by default.

I understand that your use case needs this, but doing this for everyone, by default, seems wrong. I considered whether using `.Values.agent.serviceAccount.create` makes sense, but it doesn't seem to. The values for that service account and its corresponding role will depend on what permissions that agent needs, and what the calls to `kubectl` and `helm` need. Right? Am I understanding this wrong?

I think the rest of the change (the line `{{- if .Values.agent.serviceAccountName }}` and the one following it) is fine.

Reply:
> I understand that your use case needs this, but doing this for everyone, by default, seems wrong. I considered whether using .Values.agent.serviceAccount.create makes sense, but it doesn't seem to. The values for that service account and its corresponding role will depend on what permissions that agent needs, and what the calls to kubectl and helm need. Right? Am I understanding this wrong?

How about just asking the user to provide `.Values.agent.serviceAccountName` (default value = default sa for that namespace) then? I think that's what you are suggesting, @arvindsv ? The user should manually create the static agent's service account, the relevant role and role binding depending on what their needs are. For @MPV 's case, `.Values.agent.serviceAccountName` can be equal to the name of the service account that we create for the server and elastic agents during `helm install`? This will involve a lot of setup for someone who wants to use static agents and don't want to have `.Values.agent.serviceAccountName` equal to our SA.

Reply:
@varshavaradarajan I'd think a new value such as  `.Values.agent.serviceAccount.reuse` (default = `false`) could end up reusing the server-side newly created SA. That'll allow @MPV's case, without making it mandatory in the default case.

In the default case, static agents won't have any service account associated with them (as is current behavior).

#### Case 1: A server which can create agent pods on demand.

Use:

```
serviceaccount.create = true (is already default)
```

#### Case 2: A server which can create agent pods on demand, and agents can run helm/kubectl using the same SA as that of the server.

Use:

```
serviceaccount.create = true (is already default)
agent.serviceAccount.reuse = true
```

#### Case 3: A server which can create agent pods on demand, and agents can run helm/kubectl using a different SA than that of the server.

```
serviceaccount.create = true (is already default)
agent.serviceAccount.reuse = false (is the default)
agent.serviceAccountName = some-previously-created-SA-with-right-roles-etc
```


> This will involve a lot of setup for someone who wants to use static agents and don't want to have .Values.agent.serviceAccountName equal to our SA.

Of course. They want to use a service account which is not the one we create. So, they'll need to create it and then associate it to the agents using the right property.